packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/inspections/AbstractMongoDbInspectionBridge.kt

```diff
@@ -65,21 +72,34 @@ abstract class AbstractMongoDbInspectionBridge(
                         if (fileInExpression == null || fileInExpression.virtualFile == null) {
                             inspection.visitMongoDbQuery(null, holder, cachedValue!!.value, dialect.formatter)
                         } else {
+                            val cachedQuery = cachedValue!!.value
                             val dataSource =
                                 MongoDbVirtualFileDataSourceProvider().getDataSource(
                                     expression.project,
                                     fileInExpression.virtualFile,
                                 )
+
                             inspection.visitMongoDbQuery(
                                 dataSource?.localDataSource,
                                 holder,
-                                cachedValue!!.value,
+                                queryWithCollectionReference(cachedQuery, fileInExpression.virtualFile),
                                 dialect.formatter,
                             )
                         }
                     }
                 }
             }
+
+            private fun queryWithCollectionReference(query: Node<PsiElement>, virtualFile: VirtualFile) =
+                MongoDbVirtualFileDataSourceProvider().getDatabase(
+                    virtualFile,
+                )?.let { database ->
+                    query.copy { component ->
```

Comment:
Looks good! I would however put this in the model, in a method, that encapsulates this logic, and then call the method here. A method like overwriteQueryTargetDatabase or by the likes.

Reply:
Actually part of it is what the query.copy is doing. I was trying to keep that a little generic for the purpose of aligning with the definition of copy method. Or do you mean to have this method (queryWithCollectionReference) entirely on the Node class?

Reply:
I tried something in [0276ed0](https://github.com/mongodb-js/intellij/pull/39/commits/0276ed0291e36242dbb6f65700db2701a90acfde). Was trying to keep the method pure hence keeping it static and still doing the copy